### PR TITLE
Add download functionality to DAM folders

### PR DESCRIPTION
**Pull request by jomunker**
This PR adds the functionality to download folders recursively in the DAM as ZIP.

## Video Recordings
https://github.com/vivid-planet/comet/assets/56400587/743ad1e5-b631-454f-bddb-99b919b2b12c

https://github.com/vivid-planet/comet/assets/56400587/3d90a1d3-b827-40e7-9510-4e8105b13ed8

### CODE REVIEW

File: packages/api/cms-api/src/dam/files/folders.controller.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+@Controller("dam/folders")
+export class FoldersController {
+    constructor(private readonly foldersService: FoldersService) {}
+
+    @Get("/:folderId/zip")
+    async createZip(@Param("folderId") folderId: string, @Res() res: Response): Promise<void> {
+        const folder = await this.foldersService.findOneById(folderId);
+        const folderName = folder?.name ? folder.name : "folder";
+        console.log(`starting zip generation for folder ${folderName} with id: ${folderId}`);
```
Comment (johnnyomair) [inline packages/api/cms-api/src/dam/files/folders.controller.ts:14]:
```suggestion
```

File: packages/api/cms-api/src/dam/files/folders.controller.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+import { FoldersService } from "./folders.service";
+
+@Controller("dam/folders")
+export class FoldersController {
+    constructor(private readonly foldersService: FoldersService) {}
+
+    @Get("/:folderId/zip")
+    async createZip(@Param("folderId") folderId: string, @Res() res: Response): Promise<void> {
+        const folder = await this.foldersService.findOneById(folderId);
```
Comment (johnnyomair) [inline packages/api/cms-api/src/dam/files/folders.controller.ts:12]: This should fail if the folder can't be found. You'd then won't need the fallback values for the folder name

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx [hunk 3]
```diff
@@ -4,30 +4,38 @@ import * as React from "react";
...
         <MenuItem ref={ref} {...restMenuItemProps} {...menuItemProps}>
             {icon !== undefined && <ListItemIcon {...listItemIconProps}>{icon}</ListItemIcon>}
             {children !== undefined && <ListItemText primary={children} secondary={textSecondary} {...listItemTextProps} />}
             {Boolean(endIcon) && <EndIcon>{endIcon}</EndIcon>}
         </MenuItem>
     );
-});
+};
+
+export const RowActionsListItem = React.forwardRef(RowActionsListItemNoRef) as <MenuItemComponent extends React.ElementType>(
```
Comment (johnnyomair) [inline packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: This as cast and the additional "NoRef" component seem wrong. Could you please try to fix the typing without them?

Comment (jomunker) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: I tried really long to do it cleaner but didn't find a way. If you like I can take a look again 😅

**Comment by nsams** [comment]
have you tried that with a large number of files and large files? what are the limits this can handle?

**Comment by jomunker** [comment]
> have you tried that with a large number of files and large files? what are the limits this can handle?

Don't know actually. Do you know a good way to test this rather than brute force with a lot of files?^^

**Comment by nsams** [comment]
what's bad about brute forcing (locally)?

**Comment by jomunker** [comment]
> what's bad about brute forcing (locally)?

Nothing. Just a bit annoying to do and not that realistic for production^^ I did it anyways and it has no problems with huge files or a lot of them. It just takes a while. 3GB took like 20 seconds locally

**Comment by jomunker** [comment]
I tried to optimize the time it takes with Promise.all but it made it worse

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 3 above)
Comment (thomasdax98) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: Seems like there really is no nice way to do this (https://fettblog.eu/typescript-react-generic-forward-refs/#option-1%3A-type-assertion, https://stackoverflow.com/a/58473012)

**Review by thomasdax98 — DISMISSED** [review]

File: .gitignore [hunk 4]
```diff
@@ -9,3 +9,4 @@ lang/
 .pnp.*
 junit.xml
 .env.local
+**/.idea
```
[the remark this answers is not in the document]
Comment (jomunker) [reply .gitignore:12]: thats smart thanks 🙏🏻

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 3 above)
Comment (jomunker) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: So @johnnyomair are you alright with me leaving it like this?

**Comment by johnnyomair** [comment]
@jomunker please resolve conflicts

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 3 above)
Comment (johnnyomair) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: I'd like to try this myself once you've resolved the conflicts

**Review by johnnyomair — COMMENTED** [review]
Please merge main and resolve conflicts

File: packages/api/cms-api/src/dam/files/folders.controller.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import { Response } from "express";
+
+import { FoldersService } from "./folders.service";
+
+@Controller("dam/folders")
+export class FoldersController {
+    constructor(private readonly foldersService: FoldersService) {}
+
+    @Get("/:folderId/zip")
+    async createZip(@Param("folderId") folderId: string, @Res() res: Response): Promise<void> {
```
Comment (johnnyomair) [inline packages/api/cms-api/src/dam/files/folders.controller.ts:11]: We need to verify that the user can access the scope where the folder is in

File: packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx [hunk 6]
```diff
@@ -6,6 +6,7 @@ import { saveAs } from "file-saver";
 import * as React from "react";
 import { FormattedMessage } from "react-intl";
 
+import { useCmsBlockContext } from "../../blocks/useCmsBlockContext";
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:9]: Out of scope: @thomasdax98 it's kinda strange that the DAM uses the CMS block context internally

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 3 above)
Comment (johnnyomair) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: I've tried this locally and notice that your `component` implementation isn't correct anyway 😅 Overriding the component should add the props required by the component to the props, but we're only adding component here. I'd suggest we remove the generic altogether and add a `@ts-expect-error` instead, as it seems that MUI doesn't plan to fix the issue: https://github.com/mui/material-ui/issues/33799

File: packages/api/cms-api/src/dam/files/folders.controller.ts (hunk 5 above)
Comment (jomunker) [reply packages/api/cms-api/src/dam/files/folders.controller.ts:11]: How would I do that?

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 3 above)
Comment (jomunker) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: I don't quite get what you mean exactly with 'we're only adding component here'😅

Comment (johnnyomair) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: This: `Partial<MenuItemProps<MenuItemComponent> & { component: MenuItemComponent }>`. MUI achieves component overridability by using the `OverrideProps` helper, which also adds the component's props (for instance, `href` and `target` for `a`) to the resulting type

File: packages/api/cms-api/src/dam/files/folders.controller.ts (hunk 5 above)
Comment (johnnyomair) [reply packages/api/cms-api/src/dam/files/folders.controller.ts:11]: 1. Get the current user
2. Get the scope of the folder (if DAM has enabled scoping)
3. Verify that the user can access the scope

Similar to how it's done here: https://github.com/vivid-planet/comet/blob/main/packages/api/cms-api/src/dam/images/images.controller.ts#L97

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 3 above)
Comment (jomunker) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:36]: I just tried it and it works as you can see in the screen rec. I didn't change anything in the code.

https://github.com/vivid-planet/comet/assets/56400587/6d0c835e-3f46-4fb7-ba86-484eb20a1875

File: packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx [hunk 7]
```diff
@@ -68,6 +72,14 @@ const FolderInnerMenu = ({ folder, openMoveDialog }: FolderInnerMenuProps): Reac
                     >
                         <FormattedMessage id="comet.pages.dam.rename" defaultMessage="Rename" />
                     </RowActionsItem>
+                    <RowActionsItem<"a">
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: Can't this be inferred?

Comment (jomunker) [reply packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: from the component props?

Comment (johnnyomair) [reply packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: Yes

Comment (jomunker) [reply packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: Can you show me where and how MUI does it?

Comment (johnnyomair) [reply packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: MUI does it like this:

```tsx
<Typography  component="a" href="..." />
```

`href` and other link props are automatically inferred, there's no need to specify it in the component like this:

```tsx
<Typography<"a">  component="a" href="..." />
```

Comment (jomunker) [reply packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: I tried a long time again on friday. Don't know if it's worth to put more work into this.

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx [hunk 8]
```diff
@@ -4,30 +4,38 @@ import * as React from "react";
...
     endIcon?: React.ReactNode;
-    componentsProps?: RowActionsListItemComponentsProps;
+    componentsProps?: RowActionsListItemComponentsProps<MenuItemComponent>;
     children?: React.ReactNode;
 }
 
-export const RowActionsListItem = React.forwardRef<HTMLLIElement, RowActionsListItemProps>(function RowActionsListItem(props, ref) {
+const RowActionsListItemNoRef = <MenuItemComponent extends React.ElementType>(
+    props: RowActionsListItemProps<MenuItemComponent>,
+    ref: React.ForwardedRef<HTMLLIElement>,
```
Comment (johnnyomair) [inline packages/admin/admin/src/rowActions/RowActionsListItem.tsx:22]: This should be `MenuItemComponent`, right?

File: packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx (hunk 7 above)
Comment (johnnyomair) [reply packages/admin/cms-admin/src/dam/DataGrid/DamContextMenu.tsx:75]: Alright, let's leave it like this. Maybe we'll need to fix it in the future

File: packages/admin/admin/src/rowActions/RowActionsListItem.tsx (hunk 8 above)
Comment (jomunker) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:22]: Don't get it to work with MenuItemComponent. Should a put in `any` instead?

Comment (johnnyomair) [reply packages/admin/admin/src/rowActions/RowActionsListItem.tsx:22]: Yes

**Review by johnnyomair — APPROVED** [review]

**Comment by johnnyomair** [comment]
@jomunker please fix the pipeline. Thanks!

**Review by thomasdax98 — DISMISSED** [review]

**Comment by thomasdax98** [comment]
@johnnyomair I fixed the pipeline so we can merge this

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**